presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraTokenSplitManager.java

```diff
@@ -13,129 +13,132 @@
+        return session.executeWithSession(session -> session.getCluster().getMetadata().getTokenRanges());
+    }
+
+    private List<SizeEstimate> getSizeEstimates(String keyspaceName, String tableName)
+    {
+        Statement statement = select("range_start", "range_end", "mean_partition_size", "partitions_count")
+                .from("system", "size_estimates")
+                .where(eq("keyspace_name", keyspaceName))
+                .and(eq("table_name", tableName));
+
+        ResultSet result = session.executeWithSession(session -> session.execute(statement));
+        ImmutableList.Builder<SizeEstimate> estimates = ImmutableList.builder();
+        for (Row row : result.all()) {
+            SizeEstimate estimate = new SizeEstimate(
+                    row.getString("range_start"),
+                    row.getString("range_end"),
+                    row.getLong("mean_partition_size"),
+                    row.getLong("partitions_count"));
+            estimates.add(estimate);
         }
 
-        @Override
-        public List<TokenSplit> call()
-                throws Exception
-        {
-            ArrayList<TokenSplit> splits = new ArrayList<>();
-            List<CfSplit> subSplits = client.getSubSplits(keyspace, columnFamily, range, splitSize);
-
-            // turn the sub-ranges into InputSplits
-            List<String> endpoints = range.endpoints;
-            TokenFactory factory = partitioner.getTokenFactory();
-            for (CfSplit subSplit : subSplits) {
-                Token left = factory.fromString(subSplit.getStart_token());
-                Token right = factory.fromString(subSplit.getEnd_token());
-                Range<Token> range = new Range<>(left, right, partitioner);
-                List<Range<Token>> ranges = range.isWrapAround() ? range.unwrap() : ImmutableList.of(range);
-                for (Range<Token> subRange : ranges) {
-                    TokenSplit split = new TokenSplit(factory.toString(subRange.left), factory.toString(subRange.right), endpoints);
-
-                    splits.add(split);
-                }
-            }
-            return splits;
+        return estimates.build();
+    }
+
+    private Map<TokenRange, SizeEstimate> mapSizeEstimates(List<SizeEstimate> estimates)
+    {
+        ImmutableMap.Builder<TokenRange, SizeEstimate> map = ImmutableMap.builder();
+        for (SizeEstimate estimate : estimates) {
+            TokenRange tokenRange = createTokenRange(estimate.getRangeStart(), estimate.getRangeEnd());
+            map.put(tokenRange, estimate);
         }
+        return map.build();
+    }
+
+    private TokenRange createTokenRange(String rangeStart, String rangeEnd)
+    {
+        Token rangeStartToken = createToken(rangeStart);
+        Token rangeEndToken = createToken(rangeEnd);
+        return session.executeWithSession(session -> session.getCluster().getMetadata().newTokenRange(rangeStartToken, rangeEndToken));
```

Comment:
This is a good question. There is an option to disable Metadata at all in driver. We won't do this. 

The other side is the situation when metadata is not yet loaded. In such case we should either create splits based on the even ring separation between the Presto nodes, or throw an exception saying that "Metadata is not yet available". As far as i tested it it was available from the beginning.